library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/cache/CacheTree.java

```diff
@@ -0,0 +1,179 @@
+ */
+
+package org.quiltmc.qsl.resource.loader.impl.cache;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import org.quiltmc.qsl.resource.loader.impl.ModIoOps;
+
+/**
+ * Contains the definition of a tree structure for caching, each node represents either a directory, a file, or a missing entry.
+ * <p>
+ * This may be dangerous if someone does a lot of one-time access.
+ *
+ * @author LambdAurora
+ */
+@ApiStatus.Internal
+public final class CacheTree {
+	/**
+	 * {@return a new tree, represented as a root branch node}
+	 */
+	static Branch newTree() {
+		return new Branch(null, null);
+	}
+
+	/**
+	 * Represents a node of the tree.
+	 */
+	public abstract static class Node {
+		private final Branch parent;
+		private final String path;
+		private final EntryType type;
+
+		protected Node(Branch parent, String path, EntryType type) {
+			this.parent = parent;
+			this.path = path;
+			this.type = type;
+		}
+
+		public Branch getParent() {
+			return this.parent;
+		}
+
+		public boolean isRoot() {
+			return this.getParent() == null;
+		}
+
+		public String getPathPart() {
+			return this.path;
+		}
+
+		public String getFullPath() {
+			var list = new ArrayList<String>();
+			var elem = this;
+
+			while (!elem.isRoot()) {
+				list.add(0, elem.getPathPart());
+
+				elem = elem.getParent();
+			}
+
+			return String.join("/", list);
+		}
+
+		public EntryType getType() {
+			return this.type;
+		}
+
+		public @Nullable ResourceAccess.Entry toEntry(ModIoOps ops) {
+			if (this.type == EntryType.EMPTY) return null;
+
+			return new ResourceAccess.Entry(ops.getNormalizedPath(this.getFullPath()), this.type);
+		}
+	}
+
+	/**
+	 * Represents a branch of the tree.
+	 */
+	public static final class Branch extends Node {
+		private final Map<String, Node> nodes = new ConcurrentHashMap<>();
+
+		public Branch(Branch parent, String path) {
+			super(parent, path, EntryType.DIRECTORY);
+		}
+
+		public Branch putBranch(String name) {
+			var child = new Branch(this, name);
+			this.nodes.put(name, child);
+			return child;
+		}
+
+		public void putEmpty(String name) {
+			var child = new Leaf(this, name, EntryType.EMPTY);
+			this.nodes.put(name, child);
+		}
+
+		public Leaf putFile(String name) {
+			var child = new Leaf(this, name, EntryType.FILE);
+			this.nodes.put(name, child);
+			return child;
+		}
+
+		public @Nullable Node resolveOrCompute(ModIoOps io, String path) {
+			int firstSeparator = path.indexOf('/');
+			String childName = firstSeparator == -1 ? path : path.substring(0, firstSeparator);
+
+			Node node = this.nodes.get(childName);
+
+			if (node == null) {
+				String absolutePath = childName;
+
+				if (!this.isRoot()) {
+					absolutePath = this.getFullPath() + '/' + absolutePath;
+				}
+
+				var type = io.getEntryType(absolutePath);
+
+				switch (type) {
```

Comment:
Hmm, I wonder if using directly using the switch as the returned value and replacing its internal returns with yields would be better or not, but yeah, I see a potential for that